hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/expiry/ExpirySystem.java

```diff
@@ -0,0 +1,360 @@
+        if (expireTimeByKey.isEmpty()) {
+            return ExpiryReason.NOT_EXPIRED;
+        }
+        ExpiryMetadata expiryMetadata = getExpiryMetadataForExpiryCheck(key, expireTimeByKey);
+        return hasExpired(expiryMetadata, now, backup);
+    }
+
+    private ExpiryReason hasExpired(ExpiryMetadata expiryMetadata, long now, boolean backup) {
+        if (expiryMetadata == null) {
+            return ExpiryReason.NOT_EXPIRED;
+        }
+
+        long nextExpirationTime = backup
+                ? expiryMetadata.getExpirationTime() + expiryDelayMillis
+                : expiryMetadata.getExpirationTime();
+
+        if (nextExpirationTime > now) {
+            return ExpiryReason.NOT_EXPIRED;
+        }
+
+        ExpiryReason expiryReason = expiryMetadata.getTtl() > expiryMetadata.getMaxIdle()
+                ? ExpiryReason.IDLENESS : ExpiryReason.TTL;
+        if (backup && canPrimaryDriveExpiration
+                && expiryReason == ExpiryReason.IDLENESS) {
+            return ExpiryReason.NOT_EXPIRED;
+        }
+        return expiryReason;
+    }
+
+    public InvalidationQueue<ExpiredKey> getExpiredKeys() {
+        return expiredKeys;
+    }
+
+    @Nonnull
+    public ExpiryMetadata getExpiredMetadata(Data key) {
+        ExpiryMetadata expiryMetadata = getOrCreateExpireTimeByKeyMap(false).get(key);
+        return expiryMetadata != null ? expiryMetadata : ExpiryMetadata.NULL;
+    }
+
+    @SuppressWarnings("checkstyle:magicnumber")
+    public void evictExpiredEntries(int percentage, long now, boolean backup) {
+        Map<Data, ExpiryMetadata> expireTimeByKey = getOrCreateExpireTimeByKeyMap(false);
+        if (expireTimeByKey.isEmpty()) {
+            return;
+        }
+
+        // Find max scannable key count
+        int expirableKeysMapSize = expireTimeByKey.size();
+        int keyCountInPercentage = (int) (1D * expirableKeysMapSize * percentage / ONE_HUNDRED_PERCENT);
+        int maxScannableKeyCount = Math.max(MIN_SCANNABLE_ENTRY_COUNT, keyCountInPercentage);
+
+        scanAndEvictExpiredKeys(maxScannableKeyCount, now, backup);
+
+        accumulateOrSendExpiredKey(null);
+    }
+
+    /**
+     * Get cachedExpirationIterator or init it if it has no next entry.
+     */
+    private Iterator<Map.Entry<Data, ExpiryMetadata>> getOrInitCachedIterator() {
+        if (cachedExpirationIterator == null || !cachedExpirationIterator.hasNext()) {
+            cachedExpirationIterator = initIteratorOf(expireTimeByKey);
+        }
+
+        return cachedExpirationIterator;
+    }
+
+    private void scanAndEvictExpiredKeys(int maxScannableKeyCount, long now, boolean backup) {
+        // Scan to find expired keys.
+        long scanLoopStartNanos = System.nanoTime();
+        List expiredKeyExpiryReasonList = new ArrayList<>();
+        int scannedKeyCount = 0;
+        int expiredKeyCount = 0;
+        do {
+            Map.Entry<Data, ExpiryMetadata> entry = getOrInitCachedIterator().next();
```

Comment:
What if the iterator has not next element?

Reply:
If map is empty, we return immediately before coming that line, so there should be no problem.